### PR TITLE
ATB-113 Define Metrics and Alarms for API Gateway Access logs

### PR DESCRIPTION
**Pull request by domantas-gds**
## What?

Defined the following Metric Filter and Alarm for both the API Gateway Access CloudWatch logs:

- API Gateway 4xx status errors in /apigateway/account-mgmt-frontend/access
- API Gateway 5xx status errors in /apigateway/account-mgmt-frontend/access
- API Gateway latency threshold in /apigateway/account-mgmt-frontend/access

I have enabled alarm actions to publish to our alerting SNS topics which in turn will send notifications to our `govuk-accounts-notifications` slack channel so we can monitor the sensitivity of the new defined alarms. Our PagerDuty endpoint has not yet been configured and therefore no notifications will be sent there yet. I recommend we hold off with PagerDuty integration until we are happy with the initial alarm thresholds.

**NOTE:** This PR should not be merged before https://govukverify.atlassian.net/browse/ATB-67 has been merged and has progressed through all environments.

## Why?

To facilitate monitoring of the Authentication Account Management front-end application for server and latency related errors

## Testing

Note: Testing was complete for 4xx errors as this was simple to invoke simply by hitting the home.build.account.gov.uk/manage-your-account endpoint when not signed in. Whereas 5xx would have required additional effort to invoke a status error but the implementation is the same.
Screenshot 2022-11-14 at 12 04 15

### CODE REVIEW

**Review by divyaparameswaran — DISMISSED** [review]
LGTM

**Comment by huwd** [comment]
@domantas-gds could we rebase this one off Main?
And I think we'll need to sign all the commits too 🤔

File: deploy/template.yaml [hunk 1]
```diff
@@ -1299,6 +1320,155 @@ Resources:
...
+
+  # API Gateway Alarms
+
+  # 4xx Status Code
+  ApiGateway4xxStatusAlarm:
+    Condition: IsNotDevelopment
+    Type: AWS::CloudWatch::Alarm
+    Properties:
+      AlarmName: !Sub "${AWS::StackName}-ApiGateway4xxStatusAlarm"
+      AlarmDescription: "Trigger the alarm if errorThreshold exceeds 1% with the minimum of 10 invocations in 2 out of the last 5 minutes"
```
Comment (huwd) [inline deploy/template.yaml:1335]: This sounds great to me. I might just check in with @munhitsu, I know we've some some SLAs around errors (unless you've seen them?).

I think this is a much lower threshold than what we agreed (which is great! would rather know before we breach an SLA), but i might want to check those time windows... I've got 10 mins in my mind for somre reason.

File: deploy/template.yaml [hunk 2]
```diff
@@ -1299,6 +1320,155 @@ Resources:
...
+      ActionsEnabled: false  # disabled until we're ready to invoke alarm actions
+      AlarmActions:
+        - !Sub "{{resolve:ssm:/${AlertingStackName}/SNS/HighAlertNotificationTopic/ARN}}"
```
Comment (huwd) [inline deploy/template.yaml:1338]: Nice, are these the topics we'd been mentioning in the slack PR? The priority based topics we could use for Dev Platform stuff too?

File: deploy/template.yaml [hunk 3]
```diff
@@ -1299,6 +1320,155 @@ Resources:
...
+      OKActions:
+        - !Sub "{{resolve:ssm:/${AlertingStackName}/SNS/HighAlertNotificationTopic/ARN}}"
+      ComparisonOperator: GreaterThanOrEqualToThreshold
+      EvaluationPeriods: 5
```
Comment (huwd) [inline deploy/template.yaml:1342]: Noob question, does this basically mean 5 mins?

File: deploy/template.yaml [hunk 4]
```diff
@@ -1299,6 +1320,155 @@ Resources:
...
+      DatapointsToAlarm: 2
+      Threshold: 1
+      TreatMissingData: missing
```
Comment (huwd) [inline deploy/template.yaml:1345]: Is this what happens in the absence of data?
What does missing do?

File: deploy/template.yaml [hunk 5]
```diff
@@ -1309,4 +1479,4 @@ Outputs:
     Value: !GetAtt ApiGateway.ApiEndpoint
   CloudFrontDistributionEndpoint:
     Description: The Application URL.
-    Value: !FindInMap [EnvironmentVariables, !Ref Environment, BASEURL ]
+    Value: !FindInMap [EnvironmentVariables, !Ref Environment, BASEURL ]
```
Comment (huwd) [inline deploy/template.yaml:1482]: :) Could I be a real pain and ask for a newline at the end of this file?

**Review by huwd — COMMENTED** [review]
Looks great, Would you mind rebasing, Github seems confused about the commit history, and I'm keen to only see commits listed that are on the other side of the incident.

If possible would you also mind signing your commits during that rebase:
https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

I think we're going to move to make that a requirement shortly.

I'll ping you on slack to find out how we set these thresholds, I know @munhitsu had some SLAs for us to meet, and Alex WS started work on a grafana dashboard.

Would be keen to know if this builds off that, or is just (extremely sensible looking) thresholds you've defined?

No big deal if they diverge, but would like to figure that out before ✅

File: deploy/template.yaml (hunk 1 above)
Comment (domantas-gds) [reply deploy/template.yaml:1335]: Hi @huwd I haven't seen the SLAs so if you do get some time to check in and revert that would be great! Thanks.

File: deploy/template.yaml (hunk 2 above)
Comment (domantas-gds) [reply deploy/template.yaml:1338]: Yes that's right we have two alert based topics depending on severity, it was agreed to have a high alert SNS topic for P1/P2 alerts and a low alert SNS topic for P3/P4 type alerts such as high latency as per Slack discussion with @alex9smith [here](https://gds.slack.com/archives/C011Y5SAY3U/p1669909920448659 and given we now manage our own alerting template we can add additional SNS topic pubs/subs if and when needed.

File: deploy/template.yaml (hunk 3 above)
Comment (domantas-gds) [reply deploy/template.yaml:1342]: And line 1360

File: deploy/template.yaml (hunk 4 above)
Comment (domantas-gds) [reply deploy/template.yaml:1345]: As per documentation [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarms-and-missing-data):

If all data points in the alarm evaluation range are missing, the alarm transitions to INSUFFICIENT_DATA.

Comment (domantas-gds) [reply deploy/template.yaml:1345]: So yes that's right, the alarm goes into an INSUFFICIENT_DATA state in the absence of data for a given evaluation period.

File: deploy/template.yaml (hunk 5 above)
Comment (domantas-gds) [reply deploy/template.yaml:1482]: Sure :)

**Review by divyaparameswaran — APPROVED** [review]

**Comment by huwd** [comment]
@domantas-gds shall I merge this?

**Pull request merged**